feat: Add ESP_RETURN_ON_ERR macro (IDFGH-14787)

robotman2412 opened the pull request:
## Description
This simple macro is a safe way of returning en error code rather than aborting like `ESP_ERROR_CHECK` would. It is intended to allow future pull requests by myself and/or my friend to replace unnecessary `ESP_ERROR_CHECK`s with this new `ESP_RETURN_ON_ERR` macro.

## Related
N/A

## Testing
This does not involve changing any C files and should not affect any existing code.

---

## Checklist

Before submitting a Pull Request, please ensure the following:

- [x] 🚨 This PR does not introduce breaking changes.
- [x] All CI checks (GH Actions) pass.
- [x] Documentation is updated as needed.
- [x] Tests are updated or added as necessary.
- [x] Code is well-commented, especially in complex areas.
- [x] Git history is clean — commits are squashed to the minimum necessary.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/espressif/esp-idf?pullRequest=15525) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/espressif/esp-idf?pullRequest=15525) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/espressif/esp-idf?pullRequest=15525) it.</sub>

KonstantinKondrashov commented on components/esp_common/include/esp_err.h:
```diff
@@ -92,6 +92,18 @@ void _esp_error_check_failed_without_abort(esp_err_t rc, const char *file, int l
+ * Macro that returns the error code if an error occurs, optionally running extra code.
+ * The extra code can be placed after the function thay may error and it will only run if the error occurs.
+ */
+#define ESP_RETURN_ON_ERR(x, ...) do {        \
```
I suggest renaming it as the macros in this file start with **ESP_ERROR_CHECK_** . It could be something like ESP_ERROR_RETURN, or ESP_ERROR_CHECK_RETURN.

Additionally, it would be good to have an example of usage. Could you add in the description of the macro how to use it, something like this, I guess:

```
 * @example
 * esp_err_t get_data()
 * {
 *     ESP_ERROR_CHECK_RETURN(read_file())
 *     ESP_ERROR_CHECK_RETURN(read_file(), close_file())
 *     return ESP_OK;
 * }
```

KonstantinKondrashov reviewed (commented):
Thanks for your contribution. LGTM! Just one comment is about renaming and adding an example of usage in the macro description.

asaitov commented:
You may want to move your code to `esp_check.h` or even use an existing macro https://github.com/espressif/esp-idf/blob/94cfe394fe94af5e53da8398af5b886016e7fc11/components/esp_common/include/esp_check.h#L19-L25

robotman2412 commented:
> You may want to move your code to `esp_check.h` or even use an existing macro
> 
> https://github.com/espressif/esp-idf/blob/94cfe394fe94af5e53da8398af5b886016e7fc11/components/esp_common/include/esp_check.h#L19-L25

Part of the reason for this PR is that this new macro does not log at all, as opposed to with a menuconfig option. I would be open to moving it to that file if that would get your approval for merging, but I do think it's important to have this non-log variant.

asaitov commented:
> > You may want to move your code to `esp_check.h` or even use an existing macro
> > https://github.com/espressif/esp-idf/blob/94cfe394fe94af5e53da8398af5b886016e7fc11/components/esp_common/include/esp_check.h#L19-L25
> 
> Part of the reason for this PR is that this new macro does not log at all, as opposed to with a menuconfig option. I would be open to moving it to that file if that would get your approval for merging, but I do think it's important to have this non-log variant.

Well, my intention was to make sure you considered the existing macro. I'm fine with having both.

KonstantinKondrashov commented:
Hi @robotman2412! I am ok with merging it when it is updated (rename or move to esp_check.h). Thanks.

robotman2412 commented:
@KonstantinKondrashov Hi Konstantin (and sorry for the delay). I've renamed the macro and added a simple example. Please let me know if you want any additional changes.
